src/main/java/com/flightstats/hub/filter/MetricsRequestFilter.java

```diff
@@ -49,57 +39,42 @@ public void finalStats() {
                 return;
             }
 
-            ContainerRequestContext request = requestState.getRequest();
-            long time = System.currentTimeMillis() - requestState.getStart();
-            String endpoint = getRequestTemplate(request);
-            boolean isInternal = endpoint.startsWith("/internal");
-
-            Map<String, String> tags = new HashMap<>();
-            tags.put("method", request.getMethod());
-            tags.put("call", tags.get("method") + endpoint);
-
-            String channel = RequestUtils.getChannelName(request);
-            String tag = RequestUtils.getTag(request);
-            boolean isChannel = true;
-            if (!isBlank(channel)) {
-                tags.put("channel", channel);
-            }
-            else if (!isBlank(tag)) {
-                tags.put("channel", "tag/" + tag);
-            } else {
-                isChannel = false;
-            }
-
-            if (isBlank(endpoint)) {
-                log.trace("no endpoint, path: {}", request.getUriInfo().getPath());
-            } else if (tags.get("call").endsWith("/shutdown")) {
-                log.info("call to shutdown, ignoring statsd time {}", time);
-            } else {
-                String[] tagArray = getTagArray(tags);
-                String metric = isInternal
-                        ? ( isChannel ? "internal.channel" : "internal.nonchannel" )
-                        : ( isChannel ? "api.channel" : "api.nonchannel" );
-
-                if (!statsdFilter.isTestChannel(channel) && !statsdFilter.isIgnoredRequestMetric(metric)) {
-                    log.trace("statsdReporter data sent: {}", Arrays.toString(tagArray));
-                    statsdReporter.time("request." + metric, requestState.getStart(), tagArray);
-                }
-            }
-            log.trace("request {}, time: {}", tags.get("endpoint"), time);
-            int returnCode = requestState.getResponse().getStatus();
-            if (returnCode > 400 && returnCode != 404) {
-                tags.put("errorCode", String.valueOf(returnCode));
-                String[] tagArray = getTagArray(tags, "errorCode", "call", "channel");
-                log.trace("data sent: {}", Arrays.toString(tagArray));
-                statsdReporter.count("errors", 1, tagArray);
-            }
+            reportTime(requestState.getRequestMetric(), requestState.getStart());
+            reportError(requestState);
         } catch (Exception e) {
             log.error("metrics request error", e);
         } finally {
             threadLocal.remove();
         }
     }
 
+    @VisibleForTesting
+    void reportTime(RequestMetric metric, long startTime) {
+        long time = System.currentTimeMillis() - startTime;
+        log.trace("request {}, time: {}", metric.getTags().get("endpoint"), time);
+        if (!metric.shouldReport(statsdFilter.getRequestMetricsToIgnore(), statsdFilter::isTestChannel)) {
```

Comment:
The `statsdFilter::isTestChannel` logic appears to be in the `StatsdReporter` class itself right before the metric is actually reported, so it can probably be removed here